augur/filter/subsample.py

```diff
@@ -297,6 +299,69 @@ def get_probabilistic_group_sizes(groups, target_group_size, random_seed=None):
     return max_sizes_per_group
 
 
+def get_weighted_group_sizes(groups, group_by, weights_file, target_total_size):
+    """Return group sizes based on weights defined in ``weights_file``.
+
+    Returns
+    -------
+    dict :
+        Mapping between groups (combinations of grouping column values in tuple
+        form) to group sizes
+    """
+    weights = pd.read_csv(weights_file, delimiter='\t')
+
+    weighted_columns = get_weighted_columns(weights_file)
+
+    # Allow other columns in group_by to be equally weighted (uniform sampling)
+    unweighted_columns = list(set(group_by) - set(weighted_columns))
+
+    if unweighted_columns:
+        # Augment the weights DataFrame with equal weighting for unweighted columns
+
+        # 1. Get unique values for each unweighted column
+        values_for_unweighted_columns = defaultdict(set)
+        for group in groups:
+            # NOTE: The ordering of entries in `groups` corresponds to the column
+            # names in `group_by`. but only because `get_groups_for_subsampling`
+            # conveniently retains the order. This could be more tightly coupled,
+            # but it works.
+            column_to_value_map = dict(zip(group_by, group))
+            for column in unweighted_columns:
+                values_for_unweighted_columns[column].add(column_to_value_map[column])
+
+        # 2. Create a DataFrame for all permutations of values in unweighted columns
+        lists = [list(values_for_unweighted_columns[column]) for column in unweighted_columns]
+        unweighted_permutations = pd.DataFrame(list(itertools.product(*lists)), columns=unweighted_columns)
+
+        # 3. Merge with the weights DataFrame so it has all grouping columns
+        weights = pd.merge(unweighted_permutations, weights, how='cross')
+
+    # Drop any groups that don't appear in metadata
```

Comment:
I think this is ok... initially I didn't think this was the right move but using the following example I think it's probably fine. But other people should check this bit!

Example: we want to sample twice as many location A sequences (cf. location B) for each year, but location A is missing data from one year:

```console
$ weights.tsv
year	location	weight
2001	A	2
2001	B	1
2002	A	2
2002	B	1

$ metadata.tsv
strain	date	location
SEQ3	2001	A
SEQ4	2001	B
SEQ5	2002	B
```

Targeting 100 sequences, if we drop groups missing from metadata (this PR, as it stands) we'll target `group_sizes {(2001, 'A'): 50, (2001, 'B'): 25, (2002, 'B'): 25}` and (assuming there's enough entries in the metadata) we'll get that many sequences - note that we now have a 50-50 mix of A vs B samples.

If we don't drop groups, then we'll target `group_sizes {(2001, 'A'): 33, (2001, 'B'): 16, (2002, 'A'): 33, (2002, 'B'): 16}`. The `(2002, 'A')` group will get 0 sequences, so we'll still get a 50-50 mix of A-B samples, just less of them.

Reply:
Assuming weight files are constructed by data sources spanning the full range of possible group-by values (e.g. global population sizes, case counts from all time), dropping groups will be very common and we should make sure we get this right. See https://github.com/nextstrain/ncov/pull/1106 as a real-world example.